xarray/core/variable.py

```diff
@@ -257,7 +256,10 @@ def _as_array_or_item(data):
 
     TODO: remove this (replace with np.asarray) once these issues are fixed
     """
-    data = np.asarray(data)
+    if isinstance(data, cupy_array_type):
+        data = data.get()
+    else:
+        data = np.asarray(data)
```

Comment:
I don't think you need to do this, simply try to avoid having this function called. Right now, there are shortcuts for objects defining `__array_function__`, so I think we simply need to make those available to `cupy` objects.

Reply:
Cupy does support ndarray, however they have intentionally broken it so that `np.asarray(cupy_array)` fails. This is because moving data back from the GPU is expensive. They instead broke the array API and added `.get` to force users to be sexplicit, like we are here.

Reply:
I must confess this change slipped into this PR accidentally. I got over-excited with my `git add` command. I am happy to put this into a separate PR.

Reply:
Given the conversations in  #3245 I think the best solution here is to try andmove this functionality into an `as_numpy` call. Then handle all the special casing in one place.

In the interest of speed and progress I propose that we merge this now, and then I'll look to replace it in my next PR.